TechnicalDocuments/Proposals/BuildersToFunctions.md

```diff
@@ -0,0 +1,161 @@
+# 13. Replace Builders with Free Functions
+
+* Author(s): Rui Peres
+* Review Manager: Giorgos Tsiapaliokas
+
+## Introduction
+
+The goal of this proposal is to replace the concept of a Builder as a struct to Builder as a "namespaced" static function. This could be in the following shape:
+
+```swift
+enum Builder {}
+
+extension Builder  {
+    enum Alert {
+            static func makeAlert(for error: CoreError, dismiss: (() -> Void)?) -> UIAlertController
+            static func makeAlert(for message: String, title: String?, dismiss: (() -> Void)?) -> UIAlertController
+            static func makeAlert(for error: AlertError,
+                                  primaryAction: UIAlertAction,
+                                  secondaryAction: UIAlertAction) -> UIAlertController
+    }
+}
+```
+
+This would allow us to remove the protocol `AlertBuilder` and remove all the compliances across the codebase. From a usage point of view:
+
+```swift
+final class AppointmentListFlowController {
+    let primary: Flow
+    let secondary: Flow
+
+    init(primary: Flow, secondary: Flow) {
+        self.primary = primary
+        self.secondary = secondary
+    }
+
+    func handle(_ route: AppointmentListViewModel.Route) {
+        switch route {
+        case let .showAlert(error):
+            Builder.Alert.makeAlert(for: error, dismiss: secondary.dismiss)
+                |> secondary.present
+        }
+    }
+}
+```
+
+I will provide further examples in the proposal
+
+## Motivation
+
+There are a couple of reasons for this change:
+
+1. It was decided that DI via `Current` would happen at the builder level. This means that the requirement to inject dependencies via the initializer, in builders, ceases to exist. This would: remove protocols, clean FlowController's APIs and dissolve the concept of child builders. 
+2. It would become easier to find factory methods to construct screens. One could just type `Builder.` and Xcode would autocomplete with the available factory methods. It would also free engineers from wasting too much time finding dependencies if they are immediately accessible via `Current` and the namespaced free functions. This is not a justification for engineers to not understand the codebase and how to do things "properly". It is simply an observation of how cumbersome passing dependencies via initializer is, the raison d'être for `Current` and the overall way of creating screens (VM + VC + FC). 
+
+The first point is about simplicity and clean-up of multiple ways of creating screens, while the second point focuses on discoverability and development ergonomics.   
+
+## Proposed solution
+
+**Note:** The proposed solution assumes that a way to authenticate HTTP requests is available at `Current` level (in this case something in the shape of `MainUserSession` abstracted via `UserSession`).
```

Comment:
Is it a part of this proposal to move `MainUserSession` or any other network related instances into `Current`? How can we discuss an approach that is based on something that is not in place and that we previously explicitly agreed not to do? I'd suggest to first discuss requirements for this approach to be possible.

Reply:
It's called an assumption.

Reply:
That I agree with, but as I said this change requires a separate discussion. What will be an outcome of this proposal if the decision will be that we don't change our approach with `Current`?

Reply:
Based on a preliminary and informal assessment with a couple of members of the team, moving `MainUserSession` to `Current` seems to be desirable, especially by @daniloaliberti since it would make his work easier (with the `Current` migration). 

In any case, if we don't move `MainUserSession` to `Current`, we will park this.

Reply:
To be more precise: to move `AppConfiguration` to Current, there's the `tabs` typed as `protocol TabContentBuilder` type, which declares a `make` method with a parameter of type `AuthenticatedTabContentBuildingMaterials`.
Considering such Material (and others) can be created from `Current`, then `tabs` construction would be much simpler.

Reply:
We have SDK's to do what MainUsesession does and if we go to this route we will never consume our SDK's. 
And if there is a big issue for us then maybe this is the same for our customers and we should rethink the SDK strategy.

Reply:
The name "`MainUserSession`" is used as an umbrella name to refer to an entity that is able to make authenticated and non-authenticated requests. That entity would use the SDK.

Reply:
I think @viorel15 actually raised a good point. We may have two orthogonal goals here - one is consuming sdk, another is making user session accessible globally to enable this approach (or any other approach that would avoid passing builder instances around and injecting session into them). 
We can though make `Current` consume session object from SDK (I assume there is something like that) to solve it.

Reply:
@viorel15 Please let's stay on point. If you want to discuss the usage of `MainUserSession` and SDK, let's start the discussion somewhere else. Neither has anything to do with this proposal.

Reply:
It's fine. I understood that for this proposal it does not matter if it's main user session as it is now or refactored to include the SDK's.

Reply:
@daniloaliberti this problem (dependency of app configuration on session) seems to be a real monster here, is it agreed that moving user session in the `Current` is the solution we going forward? Because there is also a parallel discussion of refactoring MainUserSession, which may solve it in a different way (I'm not up to date with it).
This problem also makes me wonder about other requirements of `AuthenticatedTabContentBuildingMaterials`, what happens with them? And another question is do we really need that in `BabylonDependencies` and `Current` specifically? It would be better to discuss that in a different channel not to diverge this discussion.